fastplotlib/graphics/vector_field.py

```diff
@@ -0,0 +1,366 @@
+        positions: np.ndarray | Sequence[float]
+            positions of the vectors, array-like, shape must be [n, 2] or [n, 3] where n is the number of vectors.
+
+        directions: np.ndarray | Sequence[float]
+            directions of the vectors, array-like, shape must be [n, 2] or [n, 3] where n is the number of vectors.
+
+        spacing: float
+            average distance between pairs of nearest-neighbor vectors, used for scaling
+
+        color: str | pygfx.Color | Sequence[float] | np.ndarray, default "w"
+            color of the vectors
+
+        vector_shape_options: dict
+            dict with the following fields that describe the shape of the vector arrows.
+            Larger values decrease the size of each component.
+
+                * cone_radius_divisor, default 10.0
+                * cone_height_divisor, default 4.0
+                * stalk_radius_divisor, default 30.0
+                * stalk_height_divisor, default 4.0
+
+        scaling_factor: float, default 1.0
+            larger values will create larger vector arrows
+
+        **kwargs
+            passed to :class:`.Graphic`
+
+        """
+
+        super().__init__(**kwargs)
+
+        self._positions = VectorPositions(positions)
+        self._directions = VectorDirections(directions)
+
+        shape_options = dict(
+            cone_radius_divisor=10.0,
+            cone_height_divisor=4.0,
+            stalk_radius_divisor=30.0,
+            stalk_height_divisor=4.0,
+        )
+
+        if vector_shape_options is None:
+            vector_shape_options = {}
+
+        for k in vector_shape_options:
+            if k not in shape_options:
+                raise KeyError(
+                    f"valid dict fields for `vector_shape_options` are: {list(shape_options.keys())}. "
+                    f"You passed the following dict: {vector_shape_options}"
+                )
+
+        shape_options = {**shape_options, **vector_shape_options}
+
+        geometry = create_vector_geometry(spacing=spacing, color=color, **shape_options)
+        material = pygfx.MeshBasicMaterial()
+        n_vectors = self._positions.value.shape[0]
+
+        world_object = pygfx.InstancedMesh(geometry, material, n_vectors)
+
+        range_3d = np.mean(np.ptp(self._positions[:], axis=0))
+        magnitudes = (
+            np.linalg.norm(self.directions[:], axis=1, ord=2) / range_3d
+        ) * size_scaling_factor
+
+        start_rot = np.array([0, 0, 1])
+
+        for i in range(n_vectors):
+            # get quaternion to rotate existing vector direction to new direction
+            rotation = la.quat_from_vecs(start_rot, self._directions[i])
+            # get the new transform
+            transform = la.mat_compose(
+                self._positions.value[i], rotation, magnitudes[i]
+            )
+            # set the buffer
+            world_object.instance_buffer.data["matrix"][i] = transform.T
+
+        world_object.instance_buffer.update_full()
```

Comment:
@almarklein this look right to you? The start position of the vector is pointing towards the viewer, so `[0, 0, 1]`.